subsys/bluetooth/controller/hci/hci.c

```diff
@@ -566,16 +599,58 @@ static void le_rand(struct net_buf *buf, struct net_buf **evt)
 static void le_read_supp_states(struct net_buf *buf, struct net_buf **evt)
 {
 	struct bt_hci_rp_le_read_supp_states *rp;
+	u64_t states = 0;
 
 	rp = cmd_complete(evt, sizeof(*rp));
 	rp->status = 0x00;
 
+#define ST_ADV (BIT64(0)  | BIT64(1)  | BIT64(8)  | BIT64(9)  | BIT64(12) | \
+		BIT64(13) | BIT64(16) | BIT64(17) | BIT64(18) | BIT64(19) | \
+		BIT64(20) | BIT64(21))
+
+#define ST_SCA (BIT64(4)  | BIT64(5)  | BIT64(8)  | BIT64(9)  | BIT64(10) | \
+		BIT64(11) | BIT64(12) | BIT64(13) | BIT64(14) | BIT64(15) | \
+		BIT64(22) | BIT64(23) | BIT64(24) | BIT64(25) | BIT64(26) | \
+		BIT64(27) | BIT64(30) | BIT64(31))
+
+#define ST_SLA (BIT64(2)  | BIT64(3)  | BIT64(7)  | BIT64(10) | BIT64(11) | \
+		BIT64(14) | BIT64(15) | BIT64(20) | BIT64(21) | BIT64(26) | \
+		BIT64(27) | BIT64(29) | BIT64(30) | BIT64(31) | BIT64(32) | \
+		BIT64(33) | BIT64(34) | BIT64(35) | BIT64(36) | BIT64(37) | \
+		BIT64(38) | BIT64(39) | BIT64(40) | BIT64(41))
+
+#define ST_MAS (BIT64(6)  | BIT64(16) | BIT64(17) | BIT64(18) | BIT64(19) | \
+		BIT64(22) | BIT64(23) | BIT64(24) | BIT64(25) | BIT64(28) | \
+		BIT64(32) | BIT64(33) | BIT64(34) | BIT64(35) | BIT64(36) | \
+		BIT64(37) | BIT64(41))
+
+#if defined(CONFIG_BLUETOOTH_BROADCASTER)
+	states |= ST_ADV;
+#else
+	states &= ~ST_ADV;
+#endif
+#if defined(CONFIG_BLUETOOTH_OBSERVER)
+	states |= ST_SCA;
+#else
+	states &= ~ST_SCA;
+#endif
+#if defined(CONFIG_BLUETOOTH_PERIPHERAL)
+	states |= ST_SLA;
+#else
+	states &= ~ST_SLA;
+#endif
+#if defined(CONFIG_BLUETOOTH_CENTRAL)
+	states |= ST_MAS;
+#else
+	states &= ~ST_MAS;
+#endif
 	/* All states and combinations supported except:
 	 * Initiating State + Passive Scanning
 	 * Initiating State + Active Scanning
 	 */
-	/*@todo: conditionally disable states based on Kconfig variables */
-	sys_put_le64(0x000003ffff3fffff, rp->le_states);
+	states &= ~(BIT64(22) | BIT64(23));
+	BT_DBG("states: 0x%08x%08x\n", (u32_t)(states >> 32), (u32_t)(states));
```

Comment:
Minor coding style stuff: no \n necessary for BT_DBG and the () around the last parameter is unnecessary.